### PR TITLE
Bug 1812204 - Add usage growth data event

### PR DESCRIPTION
**Pull request by MatthewTighe**
Note: this cannot land without an event token added to the new Event, `UsageThreshold`. Otherwise, the event will not be correctly sent.

[bugzilla ticket](https://bugzilla.mozilla.org/show_bug.cgi?id=1812204)

### GitHub Automation
Fixes #1812204

### CODE REVIEW

**Comment by MatthewTighe** [comment]
I am marking this as ready to review so that we can align on any fundamental changes to logic. This still should not be landed until after an event token is generated and added to `UsageThreshold`, but I am waiting on DS to get that to me.

File: app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt:30]: I don't expect this to be foolproof, but concerns have been raised to stakeholders about the unreliability of using lifecycle events to achieve this. It seems like this type of solution is good enough from their perspective, but I am open to any suggestions on how we might better achieve this.

File: app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt
Comment (t-p-white) [inline app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt:19]: should we use `SETTINGS` `ONE_DAY_MS`?

File: app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt
Comment (t-p-white) [inline app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt:30]:
```suggestion
         settings.usageTimeGrowthData += elapsedTimeMillis
```

File: app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt
Comment (t-p-white) [inline app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt:42]: some clarifying parameters here might be useful? e.g.
```suggestion
        return !settings.usageTimeGrowthSent && (System.currentTimeMillis() < (installTime + dayMillis))
```

**Review by t-p-white — COMMENTED** [review]
Looks good. Just a couple of suggestions which you may or may not want to action. I'll refrain from adding an approval, as there are a few named reviewers that may have more insight in this area.

File: app/src/main/java/org/mozilla/fenix/FenixApplication.kt
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/FenixApplication.kt:263]: Instead of creating the `UsageReporter` instance for every `Application#onCreate`, maybe make the `shouldRegister` a static function and wrap the instantiation and register instead that if-check?

File: app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt:41]: Probably re-use your `MetricsStorage.getInstalledTime` from there?

File: app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt:17]: Shouldn't we be delegating the storing to the MetricsStorage? We have two components that have some overlap here now.

If there is one storage accessor we have less chance of wrongly manipulating this setting.

File: app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt:42]: See above comment, but maybe this method and the storage bits should be done on the MetricsStorage?

File: app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt:17]: We should add some basic tests to ensure we are saving the time as expected. Maybe something like this? Just an idea. 🙂 

```kotlin
@Test
fun theBestTest() {
  val startTime = System.currentTimeMillis()

  UsageRecorder.onActivityResumed()
  UsageRecorder.onActivityPaused()

  val stopTime = System.currentTimeMillis()
  
  val usageTimeGrowthData = getUsageTimeGrowthData()
  assertTrue(usageTimeGrowthData < stopTime - startTime)
}

```

File: app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt
Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt:30]: This is fine, but doesn't give a clear understanding of what is happening in the activity. Are they in onboarding a custom tab (yes, you can use Firefox as a custom tab for your first session without opening the main app), browsing?

> I don't expect this to be foolproof, but concerns have been raised to stakeholders about the unreliability of using lifecycle events to measure this.

`ActivityLifecycleCallbacks` is the best option I can think of as well.

**Comment by jonalmeida** [comment]
Please update to use the appropriate issue format for title and commit messages as well. e.g. `Bug 1812204 - Add usage growth data event`

File: app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt:42]: Yeah, I like this idea. I think I was trying to keep the dependency graph a little more simple, but your suggestion will better prepare us for a future where we end up keeping this class around. I'll rework it this afternoon

File: app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt:30]: @jonalmeida I'm not sure I fully understand your concern here. We want to measure usage across _any_ activities (even the case you provided) so I don't think we need to be concerned with what is actually happening in the activity. Is that what I could make more clear in the class documentation perhaps?

Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt:30]: Sorry, I just meant the data might not be super useful to the final party. It's more a note than a concern. Regardless, you're taking the right direction by using `ActivityLifecycleCallbacks`.

File: app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/components/metrics/UsageRecorder.kt:17]: Good call, I had skipped it to avoid spinning up an actual activity and hadn't thought that step could be skipped.

**Comment by MatthewTighe** [comment]
Updated based on suggestions above. It turned into a slight refactor as I got the sense that there wasn't a real need for UsageRecorder to exist as a standalone class. Still hopeful we will soon have an opportunity to engage in more real planning around the design of this feature. 

I left it as a separate commit in case we want to revert to a standalone class, but will squash before I land. Thanks for the suggestions!

**Review by jonalmeida — COMMENTED** [review]
I left one important comment (and a small nit) that would be good if you could clarify for me, but everything else looks good.

> I got the sense that there wasn't a real need for UsageRecorder to exist as a standalone class.

FWIW, I think it's fine to have `UsageRecorder` as a separate class since it's doing one specific thing, observing the activity callbacks and then reporting them to the storage. What you've done here works well too. 🙂

File: app/src/main/java/org/mozilla/fenix/components/metrics/MetricsMiddleware.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/components/metrics/MetricsMiddleware.kt:31]: > In the second app open, we wouldn't send this data either because now usageTimeGrowthSent is true.

`usageTimeGrowthSent` should not be true until the data is actually sent in a subsequent `onResume` where the `elapsedTimeMillis` has been cached correctly, since it otherwise won't pass the conditional set up in `shouldTrack`. 

I'm open to creating something like an `AppAction.stoppedMetricsAction`, but I'm hesitant because I have a deep distrust of work finishing during destructive lifecycle events (even caching the elapsed time during onPause here feels error prone to me).

In general I am interested in refactoring the underlying design here to be a little more functional for reasons like this - managing this state through a disk cache across multiple calls seems risky. I'm still hoping that if there is continued buy-in for this kind of client-side data processing that we will have an opportunity to engage in a more robust design and implementation cycle.

**Comment by MatthewTighe** [comment]
> FWIW, I think it's fine to have UsageRecorder as a separate class since it's doing one specific thing, observing the activity callbacks and then reporting them to the storage. What you've done here works well too. 🙂

I played with a few different things, including defining `UsageRecorder` as an interface that could be referenced as a type in the `MetricsStorage` interface, but this felt like the happiest medium to me for now.

**Comment by MatthewTighe** [comment]
Updated naming and kdocs per suggestions and responded to your comment @jonalmeida. I can confirm the event is reported correctly on the Adjust backend so I am not sure if your concern is still applicable, please let me know what you think 🙂

File: app/src/main/java/org/mozilla/fenix/components/metrics/MetricsMiddleware.kt
Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/components/metrics/MetricsMiddleware.kt:31]: Ah, I missed the other part of the `shouldTrack` check: `settings.usageTimeGrowthData > usageThresholdMillis`

Doesn't that also mean we're only sending the event on the second open in that case. Does that suffice for your use case? If so, this lgtm.

> In general I am interested in refactoring the underlying design here to be a little more functional for reasons like this - managing this state through a disk cache across multiple calls seems risky. I'm still hoping that if there is continued buy-in for this kind of client-side data processing that we will have an opportunity to engage in a more robust design and implementation cycle.

Yes, this should be the right ask eventually after the end of this experiment.

**Review by jonalmeida — APPROVED** [review]
Responded to the thread above, but I don't want to block this PR any further since it sounds like testing shows that it works as expected.

File: app/src/main/java/org/mozilla/fenix/components/metrics/MetricsMiddleware.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/components/metrics/MetricsMiddleware.kt:31]: > Doesn't that also mean we're only sending the event on the second open in that case. Does that suffice for your use case? If so, this lgtm.

It will mean that yeah, but that suffices for this use case. Thanks for being thorough!

**Comment by MatthewTighe** [comment]
@mergifyio backport releases_v110.0.0

**Pull request merged**
